ENIGMAsystem/SHELL/Universal_System/Resources/sprites.cpp

```diff
@@ -0,0 +1,258 @@
+/** Copyright (C) 2008 Josh Ventura
+*** Copyright (C) 2013 Robert B. Colton
+*** Copyright (C) 2015 Harijs Grinbergs
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#include "sprites_internal.h"
+#include "Universal_System/image_formats.h"
+#include "Universal_System/nlpo2.h"
+#include "Graphics_Systems/graphics_mandatory.h"
+#include "Widget_Systems/widgets_mandatory.h"
+
+#include <string>
+#include <iostream>
+
+using enigma::Sprite;
+using enigma::sprites;
+using enigma::nlpo2dc;
+using enigma::TexRect;
+using enigma::Color;
+using enigma::RawImage;
+
+namespace {
+
+Sprite sprite_add_helper(std::string filename, int imgnumb, bool precise, bool transparent, bool smooth, bool preload, int x_offset, int y_offset, bool mipmap) {
+  unsigned int width, height, fullwidth, fullheight;
+  unsigned char *pxdata = enigma::image_load(
+      filename, &width, &height, &fullwidth, &fullheight, &imgnumb, false);
+  
+  unsigned cellwidth = width / imgnumb;
+  Sprite ns(cellwidth, height, x_offset, y_offset);
+  ns.SetBBox(0, 0, cellwidth, height);
+
+  if (pxdata == NULL) {
+    DEBUG_MESSAGE("ERROR - Failed to append sprite to index!", MESSAGE_TYPE::M_ERROR);
+    return ns;
+  }
+
+  // If sprite transparent, set the alpha to zero for pixels that should be
+  // transparent from lower left pixel color
+  if (transparent) {
+    Color c = enigma::image_get_pixel_color(pxdata, width, height, 0, height - 1);
+    enigma::image_swap_color(pxdata, width, height, c, Color {0, 0, 0, 0});
+  }
+  
+  std::vector<RawImage> rawSubimages = enigma::image_split(pxdata, width, height, imgnumb);
+  for (const RawImage& i : rawSubimages) {
+    ns.AddSubimage(i.pxdata, i.w, i.h, ((precise) ? enigma::ct_precise : enigma::ct_bbox), i.pxdata, mipmap);
+  }
+  
+  delete[] pxdata;
+  
+  return ns;
+}
+
+}
+
+namespace enigma_user {
+
+int sprite_get_width(int sprid) {
```

Comment:
Do we have tests for any of these?

Reply:
yes I wrote tests for all these long time ago plus van wrote collision tests